### PR TITLE
conn: add more detailed documentation (WIP)

### PR DESCRIPTION
**Pull request by Lotterleben**
First attempt at trying to explain the relationship between conn, applications and network stacks a bit better. This is very much WIP and currently mainly exists for @authmillenon to shout at :)

### CODE REVIEW

File: sys/include/net/conn.h [hunk 1]
```diff
@@ -12,19 +12,55 @@
...
  * * @ref conn_udp_t (net/conn/udp.h): UDP connections
  *
+ * Each network stack must implement at least one connection type.
+ *
+ * Note that this definition gives no restriction on how a connection type should be
+ * structured for simplicity and modularity. As a result, it can't give any guarantee that an
+ * implementation keeps them compatible to each other. For example, an implementation might allow
+ * a UDP receive function to be called on a raw IPv6 connection object and even choose to do so
+ * for valid reasons (e.g. code size), but this definition does not impose this on the
+ * implementation.
```
Comment (authmillenon) [inline sys/include/net/conn.h]: I don't understand this explanation ^^. Can you elaborate?

File: sys/net/gnrc/conn/udp/gnrc_conn_udp.c [hunk 2]
```diff
@@ -10,6 +10,8 @@
  * @{
  *
  * @file
+ * @brief       gnrc implementation of the udp interface defined by net/gnrc/udp.h
```
Comment (authmillenon) [inline sys/net/gnrc/conn/udp/gnrc_conn_udp.c]: GNRC is written all-caps in prose text.

File: sys/include/net/conn.h [hunk 3]
```diff
@@ -12,19 +12,55 @@
...
+ *
+ * How To Use
+ * ==========
+ *
+ * A RIOT application uses the functions provided by one or more of the connection types
+ * headers (for example @ref conn_udp_t), regardless of the network stack it uses.
+ * The network stack used under the bonnet is specified by including the appropriate
+ * module (for example USEMODULE += gnrc_conn_udp)
+ *
+ * This allows for network stack-agnostic code on the application layer.
```
Comment (authmillenon) [inline sys/include/net/conn.h]: Not sure if the hyphenation is correct.

Comment (Lotterleben) [inline sys/include/net/conn.h]: me neither :D

File: sys/include/net/conn.h [hunk 4]
```diff
@@ -12,19 +12,55 @@
...
+ *         ^
+ *         |
+ *         v
+ *       conn
+ *         ^
+ *         |
+ *         v
+ * +---------------+
+ * | Network Stack |
+ * +---------------+
```
Comment (authmillenon) [inline sys/include/net/conn.h]: This graphic will be shredded by doxygen ;) (adding four spaces before it should fix that).

Also: trailing white-spaces are evil ;-)

File: sys/include/net/conn.h (hunk 1 above)
Comment (Lotterleben) [inline sys/include/net/conn.h]: That's from the old documentation, I didn't write it ^^ I understood it as a warning that you can't rely on the implementations being completely interchangeable (for example, their definition of conn_udp may be different)?

File: sys/net/gnrc/conn/udp/gnrc_conn_udp.c (hunk 2 above)
Comment (Lotterleben) [inline sys/net/gnrc/conn/udp/gnrc_conn_udp.c]: whoops

File: sys/include/net/conn.h (hunk 4 above)
Comment (Lotterleben) [inline sys/include/net/conn.h]: Where did I leave whitespaces? o.o sublimetext should usually prevent me from being so evil :D

File: sys/include/net/conn.h (hunk 1 above)
Comment (authmillenon) [inline sys/include/net/conn.h]: Ah yes. When I find the time I will propose a better explanation :-D

File: sys/include/net/conn.h (hunk 4 above)
Comment (authmillenon) [inline sys/include/net/conn.h]: The formatting in my files view looks a little bit weird and I thought whitespace were the culprit, but I guess it's just GitHub being weird.

untitled

File: sys/include/net/conn.h (hunk 1 above)
Comment (Lotterleben) [inline sys/include/net/conn.h]: If my interpretation was correct I can also propose one for you to judge ;)

Comment (authmillenon) [inline sys/include/net/conn.h]: > Note that there might be no relation between the different connection types.
> For simplicity and modularity this API doesn't put any restriction of the actual 
> implementation of the type. For example, one implementation might choose
> to have all connection types have a common base class or use the raw IPv6
> connection type to send e.g. UDP packets, while others will keep them
> completely seperate from each other.

**Comment by authmillenon** [comment]
@Lotterleben what is there still to do (since this PR is marked WIP)?

**Comment by Lotterleben** [comment]
There's probably a boatload of info missing that I just haven't learned yet, but yeah, It doesn't make sense to stall a PR for "maybe one day"s, does it? Do you have something important in mind that I should describe before we can declare this "good enough for now"? :D

**Comment by authmillenon** [comment]
No, I would also declare it as "good enough for now". If there is more you can always add a new PR ;-).

**Comment by authmillenon** [comment]
ACK & go, when Travis is happy

**Comment by authmillenon** [comment]
Ah wait stop, please squash first :-)

**Comment by Lotterleben** [comment]
No worries, would've thought that was implied anyways ^^

**Pull request merged**